library/psa_crypto.c

```diff
@@ -5474,6 +5474,140 @@ static psa_status_t psa_key_derivation_tls12_ecjpake_to_pms_read(
 }
 #endif
 
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_HMAC)
+static psa_status_t psa_key_derivation_pbkdf2_generate_block(
+    psa_pbkdf2_key_derivation_t *pbkdf2,
+    psa_algorithm_t prf_alg,
+    uint8_t prf_output_length,
+    psa_key_attributes_t *attributes)
+{
+    psa_status_t status;
+    psa_mac_operation_t mac_operation = PSA_MAC_OPERATION_INIT;
+    size_t mac_output_length;
+    uint8_t U_i[PSA_MAC_MAX_SIZE];
+    uint8_t U_accumulator[PSA_MAC_MAX_SIZE];
+    uint64_t i;
+    uint8_t block_counter[4];
+
+    mac_operation.is_sign = 1;
+    mac_operation.mac_size = prf_output_length;
+    MBEDTLS_PUT_UINT32_BE(pbkdf2->block_number, block_counter, 0);
+
+    status = psa_driver_wrapper_mac_sign_setup(&mac_operation,
+                                               attributes,
+                                               pbkdf2->password,
+                                               pbkdf2->password_length,
+                                               prf_alg);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_update(&mac_operation, pbkdf2->salt, pbkdf2->salt_length);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_update(&mac_operation, block_counter, 4UL);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    status = psa_mac_sign_finish(&mac_operation, U_i, sizeof(U_i),
+                                 &mac_output_length);
+    if (status != PSA_SUCCESS) {
+        goto cleanup;
+    }
+
+    if (mac_output_length != prf_output_length) {
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto cleanup;
+    }
+
+    memcpy(U_accumulator, U_i, mac_output_length);
```

Comment:
Might as well target the above output to `U_accumulator` directly and skip this `memcpy`.

Reply:
Could also skip the local accumulator buffer and just have a `U_accumulator` named pointer point at the output block buffer in the operation context. That way we can skip the below memcpy and zeroizations (probably don't need to zero out the intermediate `U_i` result).

Reply:
> Might as well target the above output to `U_accumulator` directly and skip this `memcpy`.

We would still need to copy `U_1` to intermediate buffer as the input for calculation of `U_2` is `U_1`.

Reply:
I think that the original comment still applies here.

Reply:
I missed this when I read your comment earlier: I agree that we don't need `U_accumulator`, but we do need to zero out `U_i`. When the iteration count is 1 (which is used in practice, when the “password” is actually a high-entropy secret and not a human-memorable printable string), `U_i` is the output. When the iteration count is larger, `U_i` isn't enough to directly reconstruct the output without knowing the password, but it may be used as part of a larger attack path where the adversary can submit inputs with different costs and salts.